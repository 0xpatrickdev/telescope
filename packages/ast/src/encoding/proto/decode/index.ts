import * as t from '@babel/types';
import { identifier, objectMethod } from '../../../utils';
import { ProtoType, ProtoField, getFieldsTypeName, getBaseCreateTypeFuncName } from '../types';
import { baseTypes, decode } from './utils';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement decode (${field.type} rules[${field.rule}] name[${name}])`);
}

export const decodeMethodFields = (name: string, proto: ProtoType) => {
    return Object.keys(proto.fields ?? {}).map(fieldName => {
        const field = proto.fields[fieldName];
        if (field.rule === 'repeated') {
            switch (field.type) {
                case 'string':
                    return decode.scalarArray(field.id, fieldName, baseTypes.string());
                case 'uint64':
                    return decode.scalarArray(field.id, fieldName, baseTypes.long());
                case 'int32':
                    return decode.scalarArray(field.id, fieldName, baseTypes.int32());
                case 'int64':
                    return needsImplementation(fieldName, field);
                case 'bytes':
                    return needsImplementation(fieldName, field);
                default:
                    switch (field.parsedType.type) {
                        case 'Enum':
                        case 'Type':
                            return decode.typeArray(field.id, fieldName, getFieldsTypeName(field));
                    }
                    return needsImplementation(fieldName, field);
            }

        }


        if (field.keyType) {
            // currently they all look the same for decode()
            return decode.keyHash(field.id, fieldName, name);
        }

        switch (field.type) {
            case 'string':
                return decode.string(field.id, fieldName);
            case 'uint64':
                return decode.long(field.id, fieldName);
            case 'int64':
                return decode.int64(field.id, fieldName);
            case 'double':
                return decode.double(field.id, fieldName);
            case 'bytes':
                return decode.bytes(field.id, fieldName);
            case 'bool':
                return decode.bool(field.id, fieldName);
            case 'google.protobuf.Duration':
            case 'Duration':
                return decode.duration(field.id, fieldName);
            case 'google.protobuf.Timestamp':
            case 'Timestamp':
                return decode.timestamp(field.id, fieldName);
            default:
                switch (field.parsedType.type) {
                    case 'Enum':
                        return decode.enum(field.id, fieldName);
                    case 'Type':
                        return decode.type(field.id, fieldName, getFieldsTypeName(field));
                }
                return needsImplementation(fieldName, field);
        }
    });
};

export const decodeMethod = (name: string, proto: ProtoType) => {
    return objectMethod(
        'method',
        t.identifier('decode'),
        [
            identifier('input',
                t.tsTypeAnnotation(
                    t.tsUnionType(
                        [
                            t.tsTypeReference(
                                t.tsQualifiedName(
                                    t.identifier('_m0'),
                                    t.identifier('Reader')
                                ),
                                null
                            ),
                            t.tsTypeReference(
                                t.identifier('Uint8Array')
                            )
                        ]
                    )
                ),
                false
            )
        ],
        t.blockStatement([

            /*
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
            */
            t.variableDeclaration(
                'const',
                [
                    t.variableDeclarator(
                        t.identifier('reader'),
                        t.conditionalExpression(
                            t.binaryExpression(
                                'instanceof',
                                t.identifier('input'),
                                t.memberExpression(
                                    t.identifier('_m0'),
                                    t.identifier('Reader')
                                )
                            ),
                            t.identifier('input'),
                            t.newExpression(
                                t.memberExpression(
                                    t.identifier('_m0'),
                                    t.identifier('Reader')
                                ),
                                [
                                    t.identifier('input')
                                ]
                            )
                        )
                    )
                ]
            ),

            /*
    let end = length === undefined ? reader.len : reader.pos + length;
            */

            t.variableDeclaration(
                'let',
                [
                    t.variableDeclarator(
                        t.identifier('end'),
                        t.conditionalExpression(
                            t.binaryExpression('===',
                                t.identifier('length'),
                                t.identifier('undefined')
                            ),
                            t.memberExpression(
                                t.identifier('reader'),
                                t.identifier('len')
                            ),
                            t.binaryExpression(
                                '+',
                                t.memberExpression(
                                    t.identifier('reader'),
                                    t.identifier('pos')
                                ),
                                t.identifier('length')
                            )
                        )
                    )
                ]
            ),

            /*
            
    const message = createBaseMsgJoinPool();

            */

            t.variableDeclaration(
                'const',
                [
                    t.variableDeclarator(
                        t.identifier('message'),
                        t.callExpression(
                            t.identifier(getBaseCreateTypeFuncName(name)),
                            []
                        )
                    )
                ]
            ),

            ///////////
            ///////////
            ///////////

            t.whileStatement(
                t.binaryExpression(
                    '<',
                    t.memberExpression(
                        t.identifier('reader'),
                        t.identifier('pos')
                    ),
                    t.identifier('end')
                ),
                t.blockStatement([

                    /// DECODE BODY
                    t.variableDeclaration(
                        'const',
                        [
                            t.variableDeclarator(
                                t.identifier('tag'),
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('reader'),
                                        t.identifier('uint32')
                                    ),
                                    []
                                )
                            )
                        ]
                    ),


                    t.switchStatement(
                        t.binaryExpression(
                            '>>>',
                            t.identifier('tag'),
                            t.numericLiteral(3)
                        ),
                        [


                            ...decodeMethodFields(name, proto),

                            /*
                            default:
                                    reader.skipType(tag & 7);
                                    break;
                            */
                            t.switchCase(
                                null,
                                [
                                    t.expressionStatement(
                                        t.callExpression(
                                            t.memberExpression(
                                                t.identifier('reader'),
                                                t.identifier('skipType')
                                            ),
                                            [
                                                t.binaryExpression(
                                                    '&',
                                                    t.identifier('tag'),
                                                    t.numericLiteral(7)
                                                )
                                            ]
                                        )
                                    ),
                                    t.breakStatement()
                                ]
                            )
                        ]
                    )

                ])
            ),

            // RETURN STATEMENT
            t.returnStatement(
                t.identifier('message')
            )

        ]),
        false,
        false,
        false,
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.identifier(name)
            )
        )
    )
};

