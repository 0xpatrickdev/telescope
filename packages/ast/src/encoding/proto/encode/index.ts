import * as t from '@babel/types';
import { identifier, objectMethod } from '../../../utils';
import { ProtoParseContext } from '../../context';
import { ProtoType, ProtoField } from '../types';
import { encode, arrayTypes } from './utils';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement encode (${field.type} rules[${field.rule}] name[${name}])`);
}
export interface EncodeMethod {
    typeName: string;
    context: ProtoParseContext;
    field: ProtoField;
}

export const encodeMethodFields = (context: ProtoParseContext, name: string, proto: ProtoType) => {

    return Object.keys(proto.fields ?? {}).reduce((m, fieldName) => {
        const field = {
            name: fieldName,
            ...proto.fields[fieldName]
        };

        const args: EncodeMethod = {
            typeName: name,
            context,
            field
        };

        if (field.rule === 'repeated') {
            switch (field.type) {
                case 'string':
                    // TODO double check string[]
                    return [...m, ...encode.scalarArray(args, arrayTypes.string())];
                case 'uint64':
                    return [...m, ...encode.scalarArray(args, arrayTypes.long())];
                case 'int32':
                    return [...m, ...encode.scalarArray(args, arrayTypes.int32())];
                case 'int64':
                    return needsImplementation(fieldName, field);
                case 'bytes':
                    return needsImplementation(fieldName, field);
                default:
                    switch (field.parsedType.type) {
                        case 'Enum':
                            // could be same as Type?
                            return needsImplementation(fieldName, field);
                        case 'Type':
                            return [...m, ...encode.typeArray(args)];
                    }
                    return needsImplementation(fieldName, field);
            }
        }

        if (field.keyType) {
            // currently they all look the same for encode()
            return [...m, encode.keyHash(args)];
        }

        switch (field.type) {
            case 'string':
                return [...m, encode.string(args)];
            case 'uint64':
                return [...m, encode.long(args)];
            case 'double':
                return [...m, encode.double(args)];
            case 'int64':
                return [...m, encode.int64(args)];
            case 'bool':
                return [...m, encode.bool(args)];
            case 'bytes':
                return [...m, encode.bytes(args)];
            case 'Duration':
            case 'google.protobuf.Duration':
                return [...m, encode.duration(args)];
            case 'Timestamp':
            case 'google.protobuf.Timestamp':
                return [...m, encode.timestamp(args)];
            default:
                switch (field.parsedType.type) {
                    case 'Enum':
                        return [...m, encode.enum(args)];
                    case 'Type':
                        return [...m, encode.type(args)];
                }
                return needsImplementation(fieldName, field);
        }
    }, []);
};

export const encodeMethod = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const fields = encodeMethodFields(context, name, proto);
    const body = [
        ...fields,
        /* RETURN writer */
        t.returnStatement(
            t.identifier('writer')
        )
    ];

    try {
        t.blockStatement(body);
    } catch (e) {
        console.log(body);
        throw e;
    }

    return objectMethod(
        'method',
        t.identifier('encode'),
        [
            // args

            identifier('message', t.tsTypeAnnotation(
                t.tsTypeReference(
                    t.identifier(name)
                )
            ), false),

            t.assignmentPattern(
                t.identifier('writer'),
                t.callExpression(
                    t.memberExpression(
                        t.memberExpression(
                            t.identifier('_m0'),
                            t.identifier('Writer')
                        ),
                        t.identifier('create')
                    ),
                    [

                    ]
                )
            )
        ],

        // body 
        t.blockStatement(body),
        false,
        false,
        false,
        // return type
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.tsQualifiedName(
                    t.identifier('_m0'),
                    t.identifier('Writer')
                )
            )
        )
    )
};
