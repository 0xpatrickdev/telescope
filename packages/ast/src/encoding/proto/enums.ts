import * as t from '@babel/types';
import { getEnumFromJsonName, getEnumToJsonName } from './types';
import { identifier, tsEnumMember, functionDeclaration } from '../../utils';

import { ProtoEnum } from './types';

const getEnumValues = (proto: ProtoEnum) => {
    const enums = Object.keys(proto.values).map(key => {
        const e = {
            name: key,
            comment: null,
            value: null
        };
        e.value = proto.values[key];
        if (proto.comments[key]) {
            e.comment = proto.comments[key];
        }
        return e;
    });
    return enums;
}

export const createProtoEnum = (name: string, proto: ProtoEnum) => {
    const enums = getEnumValues(proto);
    const values = enums.map(e => {
        return tsEnumMember(
            t.identifier(e.name),
            t.numericLiteral(e.value),
            e.comment ? [{
                type: 'CommentBlock',
                value: e.comment
            }] : []
        );
    })
    return t.exportNamedDeclaration(
        t.tsEnumDeclaration(
            t.identifier(name),
            [
                ...values,
                // default
                tsEnumMember(
                    t.identifier('UNRECOGNIZED'),
                    t.unaryExpression('-', t.numericLiteral(1)),
                    null
                ),
            ]
        )
    )
};

export const createProtoEnumFromJSON = (name: string, proto: ProtoEnum) => {

    const enums = getEnumValues(proto);
    const switches = enums.map(e => {
        return t.switchCase(t.numericLiteral(e.value), []),
            t.switchCase(t.stringLiteral(e.name), [
                t.returnStatement(t.memberExpression(
                    t.identifier(name),
                    t.identifier(e.name)
                ))
            ]);
    });

    return t.exportNamedDeclaration(
        functionDeclaration(
            t.identifier(getEnumFromJsonName(name)),
            [
                identifier('object', t.tsTypeAnnotation(t.tsAnyKeyword()))
            ],
            t.blockStatement([
                t.switchStatement(
                    t.identifier('object'),
                    [
                        ...switches,
                        // default
                        t.switchCase(t.unaryExpression('-', t.numericLiteral(1)), []),
                        t.switchCase(t.stringLiteral('UNRECOGNIZED'), []),
                        t.switchCase(
                            null, [
                            t.returnStatement(t.memberExpression(
                                t.identifier(name),
                                t.identifier('UNRECOGNIZED')
                            ))
                        ])
                    ]
                )
            ]),
            false,
            false,
            t.tsTypeAnnotation(t.tsTypeReference(
                t.identifier(name)
            ))
        )
    )
};

export const createProtoEnumToJSON = (name: string, proto: ProtoEnum) => {

    const enums = getEnumValues(proto);
    const switches = enums.map(e => {
        return t.switchCase(
            t.memberExpression(
                t.identifier(name),
                t.identifier(e.name)
            ),
            [
                t.returnStatement(
                    t.stringLiteral(e.name)
                )
            ]
        )
    });

    return t.exportNamedDeclaration(
        functionDeclaration(
            t.identifier(getEnumToJsonName(name)),
            [
                identifier('object', t.tsTypeAnnotation(
                    t.tsTypeReference(t.identifier(name))
                ))
            ],
            t.blockStatement([
                t.switchStatement(
                    t.identifier('object'),
                    [
                        ...switches,
                        // default
                        t.switchCase(
                            null,
                            [
                                t.returnStatement(
                                    t.stringLiteral('UNKNOWN')
                                )
                            ]
                        )
                    ]
                )
            ]),
            false,
            false,
            t.tsTypeAnnotation(
                t.tsStringKeyword()
            )
        )
    )
};
