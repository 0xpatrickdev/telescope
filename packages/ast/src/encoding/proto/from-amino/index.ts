import * as t from '@babel/types';
import { ProtoType, ProtoField } from '@osmonauts/types';
import { pascal } from 'case';
import { getFieldOptionality, getFieldOptionalityForDefaults, getOneOfs } from '..';
import { identifier, objectMethod } from '../../../utils';
import { ProtoParseContext } from '../../context';
import { fromAminoJSON, arrayTypes } from './utils';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement fromAminoJSON (${field.type} rules[${field.rule}] name[${name}])`);
}
export interface FromAminoJSONMethod {
    context: ProtoParseContext;
    field: ProtoField;
    isOptional: boolean;
}

export const fromAminoJSONMethodFields = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const oneOfs = getOneOfs(proto);
    const fields = Object.keys(proto.fields ?? {}).map(fieldName => {
        const field = {
            name: fieldName,
            ...proto.fields[fieldName]
        };

        const isOneOf = oneOfs.includes(fieldName);
        const isOptional = getFieldOptionalityForDefaults(context, field, isOneOf);

        const args: FromAminoJSONMethod = {
            context,
            field,
            isOptional
        };

        if (field.rule === 'repeated') {
            switch (field.type) {
                case 'string':
                    return fromAminoJSON.array(args, arrayTypes.string());
                case 'bytes':
                    return fromAminoJSON.array(args, arrayTypes.bytes(args));
                case 'bool':
                    return fromAminoJSON.array(args, arrayTypes.bool());
                case 'float':
                    return fromAminoJSON.array(args, arrayTypes.float());
                case 'double':
                    return fromAminoJSON.array(args, arrayTypes.double());
                case 'int32':
                    return fromAminoJSON.array(args, arrayTypes.int32());
                case 'sint32':
                    return fromAminoJSON.array(args, arrayTypes.sint32());
                case 'uint32':
                    return fromAminoJSON.array(args, arrayTypes.uint32());
                case 'fixed32':
                    return fromAminoJSON.array(args, arrayTypes.fixed32());
                case 'sfixed32':
                    return fromAminoJSON.array(args, arrayTypes.sfixed32());
                case 'int64':
                    return fromAminoJSON.array(args, arrayTypes.int64());
                case 'sint64':
                    return fromAminoJSON.array(args, arrayTypes.sint64());
                case 'uint64':
                    return fromAminoJSON.array(args, arrayTypes.uint64());
                case 'fixed64':
                    return fromAminoJSON.array(args, arrayTypes.fixed64());
                case 'sfixed64':
                    return fromAminoJSON.array(args, arrayTypes.sfixed64());
                default:
                    switch (field.parsedType.type) {
                        case 'Enum':
                            return fromAminoJSON.array(args, arrayTypes.enum(args));
                        case 'Type':
                            return fromAminoJSON.array(args, arrayTypes.type(args));
                    }
                    return needsImplementation(fieldName, field);
            }
        }

        if (field.keyType) {
            switch (field.keyType) {
                case 'string':
                case 'int64':
                case 'sint64':
                case 'uint64':
                case 'fixed64':
                case 'sfixed64':
                case 'int32':
                case 'sint32':
                case 'uint32':
                case 'fixed32':
                case 'sfixed32':
                    return fromAminoJSON.keyHash(args);
                default:
                    return needsImplementation(fieldName, field);
            }
        }

        switch (field.type) {
            case 'string':
                return fromAminoJSON.string(args);
            case 'bytes':
                return fromAminoJSON.bytes(args);
            case 'bool':
                return fromAminoJSON.bool(args);
            case 'double':
                return fromAminoJSON.double(args);
            case 'float':
                return fromAminoJSON.float(args);
            case 'int32':
                return fromAminoJSON.int32(args);
            case 'sint32':
                return fromAminoJSON.sint32(args);
            case 'uint32':
                return fromAminoJSON.uint32(args);
            case 'fixed32':
                return fromAminoJSON.fixed32(args);
            case 'sfixed32':
                return fromAminoJSON.sfixed32(args);
            case 'int64':
                return fromAminoJSON.int64(args);
            case 'sint64':
                return fromAminoJSON.sint64(args);
            case 'uint64':
                return fromAminoJSON.uint64(args);
            case 'fixed64':
                return fromAminoJSON.fixed64(args);
            case 'sfixed64':
                return fromAminoJSON.sfixed64(args);
            case 'Duration':
            case 'google.protobuf.Duration':
                return fromAminoJSON.duration(args);
            case 'Timestamp':
            case 'google.protobuf.Timestamp':
                return fromAminoJSON.timestamp(args);
            default:
                switch (field.parsedType.type) {
                    case 'Enum':
                        return fromAminoJSON.enum(args);
                    case 'Type':
                        return fromAminoJSON.type(args);
                }
                return needsImplementation(fieldName, field);
        }
    });
    return fields;
};


export const fromAminoJSONMethod = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const fields = fromAminoJSONMethodFields(context, name, proto);
    let varName = 'object';
    if (!fields.length) {
        varName = '_';
    }

    const SDKTypeName =
        [name, 'SDKType']
            .filter(Boolean).join('');

    if (proto.type === 'Type' && proto.name === 'Duration') {
        // console.log(proto);
        // we could do something special here....
    }

    return objectMethod('method',
        t.identifier('fromAmino'),
        [
            identifier(varName,
                t.tsTypeAnnotation(
                    t.tsTypeReference(
                        t.identifier(SDKTypeName)
                    )
                ),
                false
            )

        ],
        t.blockStatement(
            [
                // t.variableDeclaration('const', [
                //     t.variableDeclarator(
                //         t.identifier('hoistedAsts'),
                //         t.nullLiteral()
                //     )
                // ]),

                t.returnStatement(
                    t.objectExpression(fields)
                )
            ]
        ),
        false,
        false,
        false,
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.identifier(name)
            )
        )
    )
};
