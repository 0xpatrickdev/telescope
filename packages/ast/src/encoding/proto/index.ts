export * from './decode';
export * from './encode';
export * from './enums';
export * from './types';
export * from './from-json';
export * from './from-sdk';
export * from './from-amino';
export * from './from-partial';
export * from './to-json';
export * from './to-sdk';
export * from './to-amino';
export * from './interface';