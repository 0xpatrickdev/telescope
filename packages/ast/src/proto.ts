import * as t from '@babel/types';
import { identifier } from './utils';

const NATIVE_TYPES = [
    'string',
    'int32',
    'uint32',
    'int64',
    'uint64',
    'double',
    'bytes',
    'bool',
]

export const getTSTypeFromProtoType = (type) => {
    switch (type) {
        case 'string':
            return t.tsStringKeyword();
        case 'int32':
        case 'uint32':
        case 'double':
            return t.tsNumberKeyword();
        case 'int64':
        case 'uint64':
            return t.tsTypeReference(t.identifier('Long'))
        case 'bytes':
            return t.tsTypeReference(t.identifier('Uint8Array'));
        case 'bool':
            return t.tsBooleanKeyword();
        default:
            throw new Error('getTSTypeFromProtoType() type not found');
    };
};


export interface ProtoType {
    oneofs?: { [key: string]: { oneof: string[], comment: string | undefined } },
    fields: {
        [key: string]: ProtoField;
    },
    comment: string | undefined;
}

export interface ProtoField {
    keyType?: string;
    rule?: string;
    type: string;
    id: number;
    comment: string | undefined;
}


/*
  returns "Type | undefined"
*/
const optionalType = (type: t.TSType) => {
    return t.tsUnionType([
        type,
        t.tsUndefinedKeyword()
    ])
}

const getOneOfs = (type: ProtoType) => {
    const keys = Object.keys(type.oneofs ?? {});
    if (!keys.length) return [];
    if (keys.length !== 1) throw new Error('getOneOfs() oneofs cardinality not known!');
    return type.oneofs[keys[0]].oneof;
};

const getProtoType = (field: ProtoField, isOneOf: boolean = false) => {
    let ast: any = null;

    if (NATIVE_TYPES.includes(field.type)) {
        ast = getTSTypeFromProtoType(field.type);
    } else {
        ast = t.tsTypeReference(t.identifier(field.type));
    }

    if (field.rule === 'repeated') {
        ast = t.tsArrayType(ast);
    }

    if (field.keyType) {
        ast = t.tsUnionType([
            t.tsTypeLiteral([
                t.tsIndexSignature([
                    identifier('key', t.tsTypeAnnotation(
                        getTSTypeFromProtoType(field.keyType)
                    ))
                ],
                    t.tsTypeAnnotation(ast)
                )
            ])
        ]);
    }

    if (isOneOf) return optionalType(ast);
    return ast;
}

export const createProtoType = (name: string, proto: ProtoType) => {
    const oneOfs = getOneOfs(proto);

    return t.exportNamedDeclaration(t.tsInterfaceDeclaration(
        t.identifier(name),
        null,
        [],
        t.tsInterfaceBody(
            Object.keys(proto.fields).map(field => {
                return t.tsPropertySignature(
                    t.identifier(field),
                    t.tsTypeAnnotation(
                        getProtoType(proto.fields[field], oneOfs.includes(field))
                    )
                )
            })
        )
    ))
};