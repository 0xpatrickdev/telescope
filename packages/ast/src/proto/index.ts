export * from './types';
export * from './enums';
export * from './encoding/encoding';