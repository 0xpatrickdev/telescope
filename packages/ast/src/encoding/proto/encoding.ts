import * as t from '@babel/types';
import { ProtoType } from './types';
import { fromPartialMethod } from './from-partial';
import { decodeMethod } from './decode';
import { encodeMethod } from './encode';
import { fromJSONMethod } from './from-json';
import { toJSONMethod } from './to-json';

export const createProtoObjectWithMethods = (name: string, proto: ProtoType) => {

    const methods = [
        encodeMethod(name, proto),
        decodeMethod(name, proto),
        fromJSONMethod(name, proto),
        toJSONMethod(name, proto),
        fromPartialMethod(name, proto),
    ];

    return t.exportNamedDeclaration(
        t.variableDeclaration('const',
            [t.variableDeclarator(
                t.identifier(name),
                t.objectExpression(
                    methods
                )
            )]
        )
    )
};