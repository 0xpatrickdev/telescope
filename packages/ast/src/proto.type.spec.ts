import { createProtoType } from './proto';
import struct from '../../../__fixtures__/proto-json/google/protobuf/struct.json'

import generate from '@babel/generator';

const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}
const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

it('ListValue', async () => {
    expectCode(createProtoType('ListValue', struct.root.nested.google.nested.protobuf.nested.ListValue));
});

it('Struct', async () => {
    expectCode(createProtoType('Struct', struct.root.nested.google.nested.protobuf.nested.Struct));
});

describe('oneofs', () => {
    it('Value', async () => {
        expectCode(createProtoType('Value', struct.root.nested.google.nested.protobuf.nested.Value));
    });
});
