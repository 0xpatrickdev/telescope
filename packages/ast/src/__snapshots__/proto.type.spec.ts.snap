// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListValue 1`] = `
"export interface ListValue {
  values: Value[];
}"
`;

exports[`Struct 1`] = `
"export interface Struct {
  fields: {
    [key: string]: Value;
  };
}"
`;

exports[`authz 1`] = `
"export interface authz {
  authorization: Any;
  expiration?: Date;
}"
`;

exports[`oneofs Value 1`] = `
"export interface Value {
  nullValue?: NullValue;
  numberValue?: number;
  stringValue?: string;
  boolValue?: boolean;
  structValue?: Struct;
  listValue?: ListValue;
}"
`;
